trackintel/preprocessing/staypoints.py

```diff
@@ -169,3 +178,28 @@ def generate_locations(
     ret_loc["user_id"] = ret_loc["user_id"].astype(ret_stps["user_id"].dtype)
 
     return ret_stps, ret_loc
+
+
+def _generate_locations_per_user(df, location_id_counter, distance_metric, db, geo_col):
+    """function called after groupby: should only contain records of one user; see generate_locations() function for parameter meaning."""
+
+    user_staypoints = df
+
+    # ensuring unique labels: we assume that every pandas group must have a unique index for the first element of the group
+    location_id_counter += user_staypoints.index[0]
```

Comment:
Does this imply that the original staypoints need to have a monotonic increasing index starting from 0? As staypoints' index is user input, this might not be always the case...

A solution could be we imitate the structure of [generate_staypoints](https://github.com/mie-lab/trackintel/blob/master/trackintel/preprocessing/positionfixes.py) function: 
the function here `_generate_locations_per_user` returns location with location-staypoint pairs (the inverse mapping of `labels`), and we reassign the staypoint-location mapping after the groupby operation. 
This also means we need to write the section `### create locations as grouped staypoints` (Line [108-118](https://github.com/mie-lab/trackintel/blob/4cf96573272bbdc8e475063e3edc1c36b7e44f69/trackintel/preprocessing/staypoints.py#L108)) into the groupby operation

Reply:
Do you mean that if the input staypoints file (.csv) is shuffled, it should not affect the tests? 
I just tested with a shuffled staypoints input file. I get a crash at [this](https://github.com/mie-lab/trackintel/blob/4cf96573272bbdc8e475063e3edc1c36b7e44f69/tests/preprocessing/test_staypoints.py#L172) assertion. Need to fix that.

But the master also has the same problem, if I try to run with a shuffled staypoints.csv file, [the same](https://github.com/mie-lab/trackintel/blob/4cf96573272bbdc8e475063e3edc1c36b7e44f69/tests/preprocessing/test_staypoints.py#L172) assertion fails when running master

Reply:
I think shuffled staypoints can be reordered using `start_at`; we can add this line at the beginning of the function. What I am referring to is the index of staypoints: it could be from 0 to n-1, but could also be unordered or even like "1x32sa" (all it needs to be is unique, but not necessarily ordered)

Reply:
Added sort by `start_at` and ensured unique labels for a randomized staypoint csv with string indices as suggested ([randomised_stps_string_indices.csv](https://github.com/mie-lab/trackintel/files/6576064/index_randomised.csv)).

Reply:
7901528 should address the residual concern for user_ids being strings. I have modified it to reserve new blocks of integers for every user (block size: total number of staypoints). Later in the main function, I ensure that the location ids are continuous and the _holes_ induced due to reserving entire blocks have been removed. (test file: 
[with_strings_for_user_id.csv](https://github.com/mie-lab/trackintel/files/6604470/with_strings_for_user_id.csv))